src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -665,18 +678,21 @@ class TieredSVSIndex : public VecSimTieredIndex<DataType, float> {
         assert(this->getWriteMode() != VecSim_WriteInPlace && "InPlace mode returns early");
 
         // Async mode - add vector to the frontend index and schedule an update job if needed.
-        { // Remove vector from the backend index if it exists.
-            std::scoped_lock lock(this->mainIndexGuard);
+        if (!this->backendIndex->isMultiValue()) {
+            // Remove vector from the backend index if it exists in case of non-MULTI.
+            std::lock_guard lock(this->mainIndexGuard);
             ret -= svs_index->deleteVectors(&label, 1);
             // If main index is empty then update_threshold is trainingTriggerThreshold,
             // overwise it is 1.
+        }
+        {
+            std::shared_lock lock(this->mainIndexGuard);
             update_threshold = this->backendIndex->indexSize() == 0 ? this->trainingTriggerThreshold
                                                                     : this->updateTriggerThreshold;
         }
-        { // Add vector to the frontend index and journal.
-            std::scoped_lock lock(this->flatIndexGuard, this->journal_mutex);
+        { // Add vector to the frontend index.
+            std::lock_guard lock(this->flatIndexGuard);
             ret = std::max(ret + this->frontendIndex->addVector(blob, label), 0);
```

Comment:
What should happen if we overwrite an existing label here? Should we account for it in the `swap_journal`?